### DISCUSSION TITLE
MLE questions

### DISCUSSION THREAD

**Discussion by ksathara** [openthread/openthread#10277] [Q&A] [answered]
Hi,
We have multiple SEDs and multiple FTDs, all having the same PANID and network key. Each SED has the IPV6 of its associated or paired FTD and send info to this IPV6. Now since there are multiple FTDs, there could be multiple hops being made in this transaction, which is fine. The issue we are facing is that after an SED or couple of SEDs working like this for week or months, all of a sudden this path seems broken. None of the FTDs change their locations. There may be some turned off and back on. We have verified the SEDs are sending the data continuously. What has broken is one/more hops in between FTDs to the destination. 

1. When an SED connects to a FTD via MLE, does it evaluate the full path between it and its final destination? We think not as MLE has no routing info for application level packets.
2. Else the SED connects to whatever FTD it has access to, then who's responsibility is to deliver the message? Suppose the hop in the middle gets disconnected, the SED will not know about it. Then the last FTD has to find another path I guess. But in our case it seems not. 

We are using Nordic SDK 2.3.0. I dont know which OpenThread version is being used. When I tried version command in CLI, I get 
'OPENTHREAD/2060039bb-dirty; Zephyr; May 20 2024 14:53:50'

Cheers,
Kaushalya

**Comment by jwhui** [comment]
The commit you referenced is https://github.com/openthread/openthread/commit/2060039bb

> When an SED connects to a FTD via MLE, does it evaluate the full path between it and its final destination? We think not as MLE has no routing info for application level packets.

SEDs only look to attach to a parent. Thread is a general purpose networking layer - it supports any-to-any communication and does not optimize a path to a given destination.

> Else the SED connects to whatever FTD it has access to, then who's responsibility is to deliver the message? Suppose the hop in the middle gets disconnected, the SED will not know about it. Then the last FTD has to find another path I guess. But in our case it seems not.

Routers are responsible for maintaining paths to all other routers.

Are you able to provide packet traces and logs to provide more visibility?

**Reply by ksathara** [reply]
We only have wireshark log after we see this problem. As this happens after 2-3 weeks to couple of months intervals, it's not practical to capture wireshark logs for that long. Are the packet logs after the problem any useful to you?

I could see packets being sent from one SED that got 'seemingly disconnected from its parent'. These packets were sent to its then parent FTD. I cant trace these packets after that as there is lot of traffic from that FTD onwards. Then we shutdown all the FTDs except the one which the SED got paired to. I could see the RLOC of the SED change, indicating it joined its paired FTD. Then I could see the data sent by the SED reaching the FTD, as it should be.

Also we can recover once this happens by either power cycling the FTD or the SED.

Cheers,
Kaushalya

**Reply by ksathara** [reply]
Another question, what happens if the SED only connects to any parent and if that parent cant find another router to pass the packet? i.e. the destined FTD of the packets send by the SED. Does the parent drops connection so that the SED can rescan for another? 

Cheers,
Kaushalya

**Reply by ksathara** [reply]
Any update on this?

**Reply by jwhui** [reply]
> Another question, what happens if the SED only connects to any parent and if that parent cant find another router to pass the packet?

The router will simply drop the packet. It may be possible that the router is not able to connect to other routers due to RF environment.

>  i.e. the destined FTD of the packets send by the SED. Does the parent drops connection so that the SED can rescan for another?

No, a router will not kick the child out.

**Reply by ksathara** [reply]
Can we know the route a packet is taking to its destination from SED or from its parent or the the destination router? 

>It may be possible that the router is not able to connect to other routers due to RF environment.

What kind of RF environments would do this? I guess OpenThread network should be able to stay resilient with other 2.4G networks such as mobile and BLE.  
1. Also when I shutdown other routers except the destined router, all these SEDs starts sending data back to the destined router. So that means, even if one path is blocked, at least the direct connection is possible. So why wouldn't the router that couldn't find a path, negotiate a new path? 
2. When a router couldn't find a path for one packet, would it drop all the subsequent packets or would he try to find a path every time it receives a new packet to be sent to another router?

>No, a router will not kick the child out.

So if something like 'intermediate router failing to find a path', could create a situation like we experience? Is there any workarounds to this? The SED cant detect the the path is broken isn't it? 

Sorry for the long list of questions. We are really stuck with this issue. Any help is much appreciated.

Thanks,
Kaushalya

**Reply by Irving-cl** [reply]
> What kind of RF environments would do this? I guess OpenThread network should be able to stay resilient with other 2.4G networks such as mobile and BLE.

For example, the air time is mostly occupied by things like BLE active scan or there are a lot of wifi 2.4G traffic. Or probably there are some obstacles in the place. Note this is not a problem with Thread. This will impact all radio working on 2.4Ghz. But since the SEDs can connect to the isolated router, I think at least the environment near the isolated router is ok.

**Reply by ksathara** [reply]
Thanks. Yes I can understand a router can be monetarily disconnected due to excessive radio noise, but why doesn't it heal? When I shutdown the other routers, all the sensors connected to the destined router and started sending back data. So that means at least in that instant, there was a path to the destination. Why didnt the parent router (one the SED connected to) couldn't/wouldn't find that path? 

Cheers,
Kaushalya

**Reply by ksathara** [reply]
Anyone know a way to trace the real time routes of packets in openthread? I see there is mesh diagnostics, not sure how to use it for this purpose.

**Reply by jwhui** [reply]
Logs and packet captures are best to provide more visibility.

Without more visibility, it could be anything - including external RF conditions.

**Reply by ksathara** [reply]
You mean wireshark logs? The problem I have is my wireshark logs cannot decrypt some packets. How can you track in that scenario? I have added a packet trace. For ex. in the attached log I have a sensor - rloc 0xa017. Easily I can see it sends packets to 0xa000. But from that point onwards how can I trace? 
Thanks,
Kaushalya
[Apr-24-2.zip](https://github.com/openthread/openthread/files/15428428/Apr-24-2.zip)

**Reply by Irving-cl** [reply]
Could you provide the network key of the Thread network?

**Comment by ksathara** [comment]
Hi Li,
Can we make this private? Unfortunately I cant share the networkkey in
public domain.

Cheers,
Kaushalya

**Reply by Irving-cl** [reply]
Hi Kaushalya,

It's ok. I think the key issue here is that the problematic router somehow cannot reach to other routers. And we need to figure out why that happened. But as you mentioned, if you took away the problematic router, the SEDs can connect to the dest router and can work. If you just want to make things work, I think you can probably take away that problematic router.

To troubleshoot why the problematic router cannot reach other router, let's use wireshark. Once you get your network key, you can decrypt all the packets in wireshark. In wireshark, click "Preferences" -> "IEEE 802.15.4" -> "Edit" (near decryption key). And add your network key there. Then try to check if there are any traffic between the problematic router with other routers.

![Screenshot 2024-05-26 at 23 13 23](https://github.com/openthread/openthread/assets/12945188/206ffeaa-72fa-4fb2-a9c3-650c92bf582c)

**Reply by ksathara** [reply]
Hi Li,
Thanks, but I have already added my networkkey to my wireshark. The issue is when I start my wireshark session, the MLE packets are no more. So wireshark cannot get extended source addresses of the packets going from the SEDs to their parents like below.
![image](https://github.com/openthread/openthread/assets/20408595/dd608723-1754-4c86-b959-9a422e6b1e3c)
So I can verify the SEDs are sending packets but thats pretty much it.
However I have noticed that packets send between routers can be fully decrypted. I have to verify this again. In my data, I have the MAC of the SED (added by application). So I think  still there is a way to trace it. Let me check this further and comment. The issue is I have to wait till this occurs again which may be best case couple of weeks.

Now suppose we found the problematic router, how can we add a re-initiation ?

Cheers,
Kaushalya

**Reply by Irving-cl** [reply]
Hi Kaushalya,

Once you find the problematic router, if it's a software problem (the sw somehow entered some wrong state), you can factory-reset the device and commission the device again. But if that's caused by the environment, then I think re-initialization won't solve the problem.

**Reply by ksathara** [reply]
Hi Li,
How can a router identify that it is in a problematic state? Is this via some app level handshaking mechanism or does OpenThread have some mechanism for that?

Cheers,
Kaushalya

**Reply by jwhui** [reply]
Without more visibility into the issue, it's difficult to determine if the router is behaving as expected and, if not, what the problem may be.

**Reply by ksathara** [reply]
Hi Jonathan,

Attached is a pcap log when router 0x9400 showed this behavior. I was provisioning and un-provisioning the same SED many times and all of a sudden my provisioning started failing. The SED was not getting the provisioning reply sent by the router. The router I was provisioning was 0xb000 and the SED was connected to parent 0x9400. Sensor rloc was 0x9423.
[Wireshark logs.zip](https://github.com/user-attachments/files/15508882/Wireshark.logs.zip)

Unfortunately, I cant share the network key here in public forum. If possible please send me a private message to my email. Thanks heaps for your understanding.
Cheers,
Kaushalya

**Reply by ksathara** [reply]
Hi Jonathan, Anything you can get from the logs?
Thanks,
Kaushalya

**Reply by ksathara** [reply]
![image](https://github.com/openthread/openthread/assets/20408595/01bce279-c464-4c35-a5b9-7c2aec016151)

In the above scenario, ZM1, ZM2 and ZM3 are FTDs and S1, S2..Sn are SEDs. The ZM1 (paired) is the destination router for all S1, S2..Sn sending data. They all share the same network credentials. 
My understanding is in this case ZM1, ZM3 and S1, S2 etc will form one partition and ZM2 will be on another partition. Correct me if I am wrong.

Now suppose for whatever reason ZM3 stops routing packets to ZM1 and S1, S2. etc are made to do a MLE reattach. At this point, will S1,S2..Sn can connect to ZM2 and change their partition or will they keep to the original partition and try to connect to ZM1 or ZM3? I think the latter, but please confirm.

Cheers,
Kaushalya

**Reply by jwhui** [reply]
Sn devices will connect to any parent that provides good connectivity.

**Reply by ksathara** [reply]
oh I see. So it doesn't matter there previous partition ID is different to the partition ID of ZM2? 

But if they join ZM2, none of the Sn would be able to reach any FTD from their original partition. Wouldn't this break the meshing?

What happens if Sn connects to ZM2, and ZM2 cant find a route for the Sn packets? Would ZM2 drop the connection to Sn?

Cheers,
Kaushalya

**Reply by jwhui** [reply]
End devices may send packets to any IPv6 end point. If a Thread network is partitioned, a parent cannot make any determination on the set of destinations that a child will send packets to. Just because a parent cannot forward a given packet doesn't provide any real indication whether the child wants to send to other destinations.

**Reply by ksathara** [reply]
Thanks. That means its child responsibility to make sure it's packets get to the intended recipient and if that doesn't happen, then negotiate a different parent? 

In that context, I have tried thread detach and attach on a SED which is connected to a particular parent, but every time it seems to get the same parent. Is there a way for the SED to reject a parent?

Cheers,
Kaushalya

**Reply by jwhui** [reply]
> That means its child responsibility to make sure it's packets get to the intended recipient and if that doesn't happen, then negotiate a different parent?

Thread is not currently designed to support this use case.

**Reply by ksathara** [reply]
Thanks Jonathan, sorry about long list of questions. I couldn't find clear answers from OpenThread specification.

In that case, if a SED connects to a parent in a different partition, all its data intended for a recipient in the previous partition, will be lost. And there is no way to recover from this. Is my understanding correct?

If that is correct, then having partitions is a big problem, isnt it?

Cheers,
Kaushalya

**Reply by jwhui** [reply]
You should ensure your topology has enough router-capable devices to avoid partitions.

If your mesh is sparse enough to have partitions, there is not much redundancy and the benefits of mesh are not realized.

**Reply by ksathara** [reply]
Hi, Sorry for my delay as I was not well past week. 

Is there a possibility that a SED can do a MLE attach and  make sure it gets a 'new parent' each time? i.e. like the SED 'walk-through' each accessible parent in the SED's range until some condition is met.

**Reply by jwhui** [reply, marked as answer]
Such a capability would not be conformance to the Thread Specification. OpenThread does not implement such a capability.

**Reply by ksathara** [reply]
ok, thanks Jonathan. Much appreciated.
